File: placementprepration/tree/dails.java
import java.util.*;
class dails{
    static class edge{
        int to, weight;
        edge(int to, int weight){
            this.to=to;
            this.weight=weight;
        }
    }
    public static void dialsalgo(int N, int src, List<List<edge>>adj, int w){
        int inf=Integer.MAX_VALUE;
        int[]dist=new int [N];
        Arrays.fill(dist,inf);
        dist[src]=0;

        // made of bukset to store the node
        List<Deque<Integer>> buckets = new ArrayList<>();
        for(int i=0;i<=w*N; i++){
            buckets.add(new LinkedList<>());
        }
        // enter the source node in bucket 
        buckets.get(0).add(src);

//process the bucket

        for(int d=0;d<=w*N;d++){
            while(!buckets.get(d).isEmpty()){
                int u=buckets.get(d).poll();
                if(dist[u]<d) continue;

                for(edge e : adj.get(u)){
                    int v=e.to,weight=e.weight;
                    if(dist[v]>dist[u]+weight){
                        dist[v]=dist[u]+weight;
                        buckets.get(dist[v].add(v));
                    }
                }

            }
        }

        
        // print shortest distance 
        for(int i=0;i<N;i++){
            System.out.println("distance");
        }


    }
    public static void main(String[] args){
        int N=5;
        List<List<edge>> adj=new ArrayList<>();
        for(int i=0;i<N;i++){
            adj.add(new ArrayList<>());
        }
        adj.get(0).add(new edge(1,2));
        adj.get(1).add(new edge(2,1));
        adj.get(1).add(new edge(3,3));
        adj.get(2).add(new edge(4,4));
        adj.get(3).add(new edge(4,2));

        int maxweight=4;
        dials(N,0,adj,maxweight);
    }
}

File: placementprepration/tree/dails.py
from collections import deque

class Edge:
    def __init__(self, to, weight):
        self.to = to
        self.weight = weight


def dials_algo(N, src, adj, w):
    inf = 10 ** 18
    dist = [inf] * N
    dist[src] = 0
    buckets = [deque() for _ in range(w * N + 1)]
    buckets[0].append(src)
    for d in range(w * N + 1):
        while buckets[d]:
            u = buckets[d].popleft()
            if dist[u] < d:
                continue
            for e in adj[u]:
                v, wt = e.to, e.weight
                if dist[v] > dist[u] + wt:
                    dist[v] = dist[u] + wt
                    buckets[dist[v]].append(v)
    return dist


def main():
    N = 5
    adj = [[] for _ in range(N)]
    adj[0].append(Edge(1, 2))
    adj[1].append(Edge(2, 1))
    adj[1].append(Edge(3, 3))
    adj[2].append(Edge(4, 4))
    adj[3].append(Edge(4, 2))
    maxweight = 4
    dist = dials_algo(N, 0, adj, maxweight)
    print("distance", dist)


if __name__ == "__main__":
    main()
